src/librustc/ty/util.rs

```diff
@@ -1116,6 +1124,8 @@ fn needs_drop_raw<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
         ty::TyClosure(def_id, ref substs) => substs.upvar_tys(def_id, tcx).any(needs_drop),
 
+        ty::TyGenerator(..) => true,
```

Comment:
I guess it's a bit fragile, but I just made generator always have destructors since that is the simpler option (and I had no generator interior when this code was written).